cli/internal/helm/chart.go

```diff
@@ -0,0 +1,354 @@
+package helm
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/defenseunicorns/zarf/cli/types"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/defenseunicorns/zarf/cli/internal/k8s"
+	"github.com/defenseunicorns/zarf/cli/internal/message"
+	"github.com/defenseunicorns/zarf/cli/internal/utils"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/storage/driver"
+)
+
+type ChartOptions struct {
+	BasePath      string
+	Chart         types.ZarfChart
+	ReleaseName   string
+	ChartOverride *chart.Chart
+	ValueOverride map[string]interface{}
+	Images        []string
+}
+
+type renderer struct {
+	images     []string
+	namespaces []string
+}
+
+// InstallOrUpgradeChart performs a helm install of the given chart
+func InstallOrUpgradeChart(options ChartOptions) {
+	spinner := message.NewProgressSpinner("Processing helm chart %s:%s from %s",
+		options.Chart.Name,
+		options.Chart.Version,
+		options.Chart.Url)
+	defer spinner.Stop()
+
+	var output *release.Release
+
+	options.ReleaseName = fmt.Sprintf("zarf-%s", options.Chart.Name)
+	actionConfig, err := createActionConfig(options.Chart.Namespace)
+
+	// Setup K8s connection
+	if err != nil {
+		spinner.Fatalf(err, "Unable to initialize the K8s client")
+	}
+
+	attempt := 0
+	for {
+		attempt++
+
+		spinner.Updatef("Attempt %d of 3 to install chart", attempt)
+		histClient := action.NewHistory(actionConfig)
+		histClient.Max = 1
+
+		if attempt > 2 {
+			// On total failure try to rollback or uninstall
+			if histClient.Version > 1 {
+				spinner.Updatef("Performing chart rollback")
+				_ = rollbackChart(actionConfig, options.ReleaseName)
+			} else {
+				spinner.Updatef("Performing chart uninstall")
+				_, _ = uninstallChart(actionConfig, options.ReleaseName)
+			}
+			spinner.Errorf(nil, "Unable to complete helm chart install/upgrade")
+			break
+		}
+
+		spinner.Updatef("Checking for existing helm deployment")
+		if _, histErr := histClient.Run(options.ReleaseName); histErr == driver.ErrReleaseNotFound {
```

Comment:
I think this conditional block has a couple bugs in it. Does this look more right to you?


```
if _, histErr := ............ {
    ........
} else if histErr == driver.ErrReleaseExists {.         // check histErr instead of err and check for upgrade before just assuming failure
    upgradeChart( ... )
} else {
    fatal("unable to verify chart"
}

Reply:
Just kidding, this change breaks the `zarf init` with a nil pointer. Either way there's a bug here we should work out.